proposals/p5168.md

```diff
@@ -0,0 +1,1015 @@
+    See
+    [proposal #5087: Qualified lookup into types being defined](https://github.com/carbon-language/carbon-lang/pull/5087).
+-   We say an entity is _complete_ at the end of its definition (at the close
+    curly `}`).
+-   We say a facet type is _identified_ if all the interfaces it references are
+    declared and all of its named constraints are complete. An identified facet
+    type is associated with a known set of interfaces.
+-   An _implied constraint_ is a condition that must hold if a type is used in a
+    given type expression, which we enforce in some way other than checking that
+    the condition holds at that point. For example:
+
+    ```
+    interface Hash { ... }
+    class HashSet(T:! Hash) { ... }
+
+    // `U` must satisfy the constraints on the `T` parameter to `HashSet`.
+    // In this case, we enforce it using an implied constraint, so we don't
+    // require `U impls Hash` beforehand (say by being declared `U:! Hash`).
+    fn Contains[U:! type](needle: U, haystack: HashSet(U)) -> bool;
+    // In this case, `Contains` is equivalent to this declaration that
+    // doesn't use implied constraints:
+    fn Contains[U:! type where .Self impls Hash]
+               (needle: U, haystack: HashSet(U)) -> bool;
+
+    // In this example, the implied constraint is on something more
+    // complicated than a single symbolic parameter.
+    fn F[V:! type, W:! type](x: HashSet(Pair(V, W)));
+    // is equivalent to:
+    fn F[V:! type, W:! type where Pair(V, .Self) impls Hash]
+        (x: HashSet(Pair(V, W)));
+    ```
+
+    -   We generally require that there always be a way to explicitly declare
+        constraints so implied constraints are not needed.
+    -   Implied constraints allow declarations to be more concise, particularly
+        we hope that common requirements can be omitted since they are implied.
+    -   Implied constraints are also a tool we can choose to use to say that a
+        requirement on a parameter can be enforced at a later point. In this
+        example,
+
+        ```
+        interface B {}
+        interface A;
+        class C(T:! B);
+        fn F(U:! A, x: C(U));
+
+        interface A {
+        require Self impls B;
+        }
+        ```
+
+        the symbolic type parameter `U` has a constraint that it impls `A`, and
+        we can add an implied constraint that it also impls `B` because it is
+        used as an argument to `C`. This constraint might be redundant with a
+        requirement from the definition of `A`, which may or may not have been
+        available at the point where `F` is declared. Callers have to provide
+        arguments that satisfy the additional requirement if it turns out that
+        `A` alone is not sufficient.
+
+### Other considerations
+
+Different kinds of entities have different requirements (ignoring `extern`
+declarations):
+
+-   A generic or compile-time function must be defined before it is called.
```

Comment:
For a generic function, shouldn't it be sufficient for it to be defined later in the same file as the call?

Reply:
I guess I'm thinking about evaluation of a function at compile time, where the evaluation may be later than the definition of the call. How should I phrase that?

Reply:
If you want to get into that level of detail, then maybe something along these lines:

"A compile-time function must be defined before a call to it is evaluated. A generic function must be defined before the end of each file in which a call to it is type-checked."

Reply:
Updated.